core/src/main/java/com/linecorp/armeria/client/Http1ResponseDecoder.java

```diff
@@ -263,4 +279,47 @@ public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exceptio
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
         ctx.fireExceptionCaught(cause);
     }
+
+    void setKeepAliveHandler(ChannelHandlerContext ctx, KeepAliveHandler keepAliveHandler) {
+        this.keepAliveHandler = keepAliveHandler;
+        maybeInitializeKeepAliveHandler(ctx);
+    }
+
+    private void maybeInitializeKeepAliveHandler(ChannelHandlerContext ctx) {
+        if (keepAliveHandler != null && ctx.channel().isActive()) {
+            keepAliveHandler.initialize(ctx);
+        }
+    }
+
+    private void destroyKeepAliveHandler() {
+        if (keepAliveHandler != null) {
+            keepAliveHandler.destroy();
+        }
+    }
+
+    private void onPingRead(Object msg) {
+        if (msg instanceof LastHttpContent) {
+            onPingComplete();
+        } else {
+            assert keepAliveHandler != null;
+            keepAliveHandler.onPing();
+        }
+    }
+
+    void setPingReqId(int id) {
+        lastPingReqId = id;
+    }
+
+    boolean isPingReqId(int id) {
+        return lastPingReqId == id;
+    }
```

Comment:
What happens if a server doesn't respond to a ping? Do we send another `OPTIONS *`? I guess we should not send another `OPTIONS *` but close the connection, unlike HTTP/2.